File: client/lib/services/api/api_result.dart
class ApiResult<T> {
  final T? data;
  final String? error;
  ApiResult.success(this.data) : error = null;
  ApiResult.error(this.error) : data = null;
  bool get isSuccess => error == null;

  factory ApiResult.fromJson(Map<String, dynamic> json) {
    if (json['error'] != null) {
      return ApiResult.error(json['error'] as String);
    } else {
      return ApiResult.success(json['data'] as T?);
    }
  }
}

File: server/app/schemas/api_result.py
from typing import Generic, TypeVar, Optional

T = TypeVar('T')

class ApiResult(Generic[T]):
    def __init__(self, data: Optional[T] = None, error: Optional[str] = None):
        self.data = data
        self.error = error

    @property
    def is_success(self) -> bool:
        return self.error is None

    @classmethod
    def success(cls, data: T = None):
        return cls(data=data)

    @classmethod
    def error(cls, error: str):
        return cls(error=error)
